STORM-1280 port backtype.storm.daemon.logviewer to java

HeartSaVioR opened the pull request:
* port logviewer to Java using Dropwizard
  * follow implementation of DRPC Server
* add '/api/v1/' to the logviewer API endpoints
  * seems like it is not possible (or not easy) to serve both API endpoints and static resources in same servlet
  * reflect the change to the static resources

Please note that logviewer test should be ported to Java since ported version of Logviewer is placed to storm-webapp module. Below is TODO list:

TODO
* [x] port logviewer-test.clj to Java and delete logviewer-test.clj
* [x] delete logviewer.clj and rename package to remove 'wip'
* [x] apply the change for https://github.com/apache/storm/commit/0201ae79a5f4fc7ba7ee1c524818a29a92f140f1#diff-81a8fa5a47d77bd6cca8d98d9916f172 and rebase this to master.
* [x] debug and fix failing tests
* [x] try best to reduce checkstyle violations

I've added J2HTML and jOOL as dependencies for storm-webapp: J2HTML as a replacement for hiccup, and jOOL as Java 8 lambda helper.

Manually tested. Unit test will be covered with porting logviewer-test.clj.

Please let me know deleting logviewer.clj is more convenient to review.
Please review and comment. Thanks in advance!

HeartSaVioR commented:
Also removed unused imports from other spots. I hope this is OK.

HeartSaVioR commented:
DirectoryCleaner also can be moved to storm-webapp after porting logviewer-test.clj

HeartSaVioR commented:
UPDATE:
I also ported logviewer-test.clj to Java. Some tests are failing and I'm looking into it.
I still need to apply recent change of logviewer: 
https://github.com/apache/storm/commit/0201ae79a5f4fc7ba7ee1c524818a29a92f140f1#diff-81a8fa5a47d77bd6cca8d98d9916f172.

HeartSaVioR commented:
Applied recent change of logviewer and rebased.

HeartSaVioR commented:
Also fixed the tests.

HeartSaVioR commented:
Removed WIP though I still need to address checkstyle issues. I think check style violation doesn't hurt review process.

HeartSaVioR commented:
Test is failing with RAT issue.

```
[WARNING] Files with unapproved licenses:
  storm-webapp/src/test/resources/test-3072.log.test
  storm-webapp/src/test/resources/small-worker.log.test
  storm-webapp/src/test/resources/test-worker.log.test
  storm-webapp/src/test/resources/logviewer-search-context-tests.log.test
```

I'll address it after addressing checkstyle.

HeartSaVioR commented:
Addressed RAT and checkstyle violation.

HeartSaVioR commented on storm-buildtools/storm_checkstyle.xml:
```diff
@@ -46,6 +46,9 @@
  -->
 
 <module name = "Checker">
+    <!-- Filter out Checkstyle warnings that have been suppressed with the @SuppressWarnings annotation -->
```
I just add allowing SuppressWarnings annotation, since the test code contains long string constant. I didn't recognize that maven checkstyle doesn't check test codes, but it might be useful for specific situations anyway.

Ethanlm commented:
I did some simple tests manually in non-secure and secure storm cluster. It seems that it works well on non-secure storm cluster except some display issues. I found some issues on secure storm cluster. I am not sure if these only happen in my cluster.  Thanks.

1. "Prev" "First" etc buttons seems to mix together.  (On non-secure and secure cluster)
![screen shot 2017-07-18 at 1 38 04 pm](https://user-images.githubusercontent.com/14900612/28334105-b00ae258-6bbf-11e7-8601-10bac362862a.png)

2. Got "page not found" while querying worker.log on **secure** cluster. The file actually exists. 
![screen shot 2017-07-18 at 1 41 32 pm](https://user-images.githubusercontent.com/14900612/28334320-6c28e98a-6bc0-11e7-97ed-65e626767f35.png)

3. Got "Request failed" which querying dumps.
![screen shot 2017-07-18 at 1 42 01 pm](https://user-images.githubusercontent.com/14900612/28334413-b0546d64-6bc0-11e7-80c7-9206e0fe0c6f.png)
The error message from logviewer.log:
<img width="728" alt="screen shot 2017-07-18 at 1 42 39 pm" src="https://user-images.githubusercontent.com/14900612/28334447-d2182134-6bc0-11e7-93b7-a02ee68ea421.png">

HeartSaVioR commented:
@Ethanlm 
> 1. "Prev" "First" etc buttons seems to mix together. (On non-secure and secure cluster)

Looks like failing to load css and also js file. We don't include static resource files to the class path, so you need to build binary distribution and use it to test correctly.
(Please let me know if you tested it with binary distribution.)

Btw, it seems to be by intention to avoid adding static resources to the client side. Maybe we can move static files to storm-webapp after porting UI and Logviewer is completed, since storm-webapp module is not intended to be used (as a dependency) for Storm client. It also helps to run and test Logviewer and UI with IDE.

> 2. Got "page not found" while querying worker.log on secure cluster. The file actually exists.

If you could spend some time to help me debugging, could you check the file's permission? 

> 3. Got "Request failed" which querying dumps on secure cluster

Is it occurred only from secure cluster? Looks like actual error is not relevant to secure, so just curious.

To be honest, I didn't have a chance to deploy the patch to secured cluster, since I feel there should be  no difference between non-secure and secure cluster. I might be wrong. I'll also spin secured cluster and test it again.

Ethanlm commented:
1. I believe I tested it from binary distribution. I used "mvn clean install" to compile and "cd storm-dist/binary && mvn package" to created the binary distribution. And then I deployed it under /tmp directory on my openstack instance. I do it every time when I need to do some tests

2. Will try to do something about it.

3. It happens only on secure cluster. Non-secure cluster works fine in my tests. I could have done something wrong. It would be great if you could test it on your secure cluster.

Thanks.

revans2 commented:
I was playing around with it and I am seeing `&gt;` in the output log. i.e.

```
2017-07-19 15:00:49.413 o.a.s.d.s.Slot SLOT_6701 [INFO] STATE EMPTY msInState: 38110 -&gt; WAITING_FOR_BASIC_LOCALIZATION msInState: 1
```

It looks like we might be double escaping some things.

I'll play around with it some more, and then start looking through the code.

revans2 commented on storm-webapp/src/main/java/org/apache/storm/daemon/logviewer/webapp/LogviewerResource.java:
```diff
@@ -0,0 +1,268 @@
+        String callback = request.getParameter("callback");
+        String origin = request.getHeader("Origin");
+
+        return logviewer.listLogFiles(user, portStr != null ? Integer.parseInt(portStr) : null, topologyId, callback, origin);
```
In the original code we handled an InvalidRequestException. Do we need to do that here too?

HeartSaVioR replied:
InvalidRequestException is a checked exception. Clojure seems not check whether the exception class is checked or runtime so placing InvalidRequestException is possible, but from here it doesn't throw InvalidRequestException so can't place it.

revans2 replied:
OK Sounds good.

revans2 commented on storm-webapp/src/main/java/org/apache/storm/daemon/logviewer/webapp/LogviewerResource.java:
```diff
@@ -0,0 +1,268 @@
+        String callback = request.getParameter("callback");
+        String origin = request.getHeader("Origin");
+
+        return logviewer.listLogFiles(user, portStr != null ? Integer.parseInt(portStr) : null, topologyId, callback, origin);
```
Do we need to handle an InvalidRequestException?

HeartSaVioR replied:
Same as above.

revans2 commented on storm-webapp/src/main/java/org/apache/storm/daemon/logviewer/webapp/LogviewerResource.java:
```diff
@@ -0,0 +1,268 @@
+        String user = httpCredsHandler.getUserName(request);
+        String file = request.getParameter("file");
+        String decodedFileName = URLDecoder.decode(file);
+        return logDownloadHandler.downloadLogFile(decodedFileName, user);
```
Do we need to handle InvalidRequestException?

HeartSaVioR replied:
Same as above.

revans2 commented on storm-webapp/src/main/java/org/apache/storm/daemon/logviewer/webapp/LogviewerResource.java:
```diff
@@ -0,0 +1,268 @@
+        String user = httpCredsHandler.getUserName(request);
+        String file = request.getParameter("file");
+        String decodedFileName = URLDecoder.decode(file);
+        return logDownloadHandler.downloadDaemonLogFile(decodedFileName, user);
```
Again InvalidRequestException?

HeartSaVioR replied:
Same as above.

revans2 commented on storm-webapp/src/main/java/org/apache/storm/daemon/logviewer/webapp/LogviewerResource.java:
```diff
@@ -0,0 +1,268 @@
+        String startByteOffset = request.getParameter("start-byte-offset");
+        String searchArchived = request.getParameter("search-archived");
+        String callback = request.getParameter("callback");
+        String origin = request.getHeader("Origin");
```
Just an FYI Headers and parameters can be marked with annotations that are pulled out as parameters.  I am not sure if this is cleaner or if that would be.  Bother are rather long.

HeartSaVioR replied:
Yes agreed. I'm also not sure which is clearer, so I just applied what we did.

revans2 commented on storm-webapp/src/main/java/org/apache/storm/daemon/logviewer/handler/LogviewerLogPageHandler.java:
```diff
@@ -0,0 +1,467 @@
+            if (port == null) {
+                fileResults = new ArrayList<>();
+
+                File topoDir = new File(logRoot + Utils.FILE_PATH_SEPARATOR + topologyId);
```
nit:  I know this is an exact translation, but we could just use `new File(logRoot, topologyId)`

HeartSaVioR replied:
Yes it would be better to fix that. I'll address it.

HeartSaVioR commented on storm-webapp/src/main/java/org/apache/storm/daemon/logviewer/LogviewerServer.java:
```diff
@@ -0,0 +1,182 @@
+            //This is kind of useless...
+            meterShutdownCalls.mark();
+
+            //TODO this is causing issues...
```
Actually most of part for LogviewerServer is copy and paste of DRPCServer, and that comment was from there. Could you revisit DRPCServer and address comments?

revans2 replied:
I thought it looked familiar.  I'll file a follow on JIRA to try and clean it up once we have everything ported over to java.

HeartSaVioR commented on storm-webapp/src/main/java/org/apache/storm/daemon/logviewer/LogviewerServer.java:
```diff
@@ -0,0 +1,182 @@
+     * @return The port the HTTP server is listening on. Not available until {@link #start() } has run.
+     */
+    public int getHttpServerPort() {
+        assert httpServer.getConnectors().length == 1;
```
Same here: this is from copy and paste, and even DRPCServer it seems to be only used for unit test. We could just remove the method and re-add if needed.

HeartSaVioR commented on storm-webapp/src/main/java/org/apache/storm/daemon/logviewer/LogviewerServer.java:
```diff
@@ -0,0 +1,182 @@
+            final Boolean httpsWantClientAuth = (Boolean) (conf.get(DaemonConfig.LOGVIEWER_HTTPS_WANT_CLIENT_AUTH));
+            final Boolean httpsNeedClientAuth = (Boolean) (conf.get(DaemonConfig.LOGVIEWER_HTTPS_NEED_CLIENT_AUTH));
+
+            //TODO a better way to do this would be great.
```
The comment was copied and pasted from DRPCServer.

While porting I thought about how to do it better, but no idea given that LogviewerApplication instance is generated dynamically. It may be possible to not passing config map and re-read storm configuration in LogviewerApplication, but it is duplicated effort and we will not be able to inject other config map in unit test. 

Maybe we could just remove the TODO comment.

HeartSaVioR commented on storm-core/src/clj/org/apache/storm/ui/core.clj:
```diff
@@ -134,12 +134,12 @@
 (defn logviewer-link [host fname secure?]
   (if (and secure? (*STORM-CONF* LOGVIEWER-HTTPS-PORT))
-    (UIHelpers/urlFormat "https://%s:%s/log?file=%s"
+    (UIHelpers/urlFormat "https://%s:%s/api/v1/log?file=%s"
```
Yes agreed. There're some responses returning HTML page instead of JSON, so we may need to change it first and try to document it.

revans2 commented:
Overall it looks good.  At least as an accurate translation of the clojure code.  I want to play around with it some more, and also try to understand what is happening with some of the escaped charters (just to see if it was happening before too).

revans2 commented:
I did a bunch of testing and I am +1 on merging this in.  There is still the issue with double escaping the contents returned for a file, but I think we can fix that after if we want.

HeartSaVioR commented:
@revans2 Just to check: did you play around secured cluster as well? I still didn't test this on secured cluster, just now set up secured cluster and plan to set up Storm and test.

Ethanlm commented:
@HeartSaVioR An update about _"(2)Got "page not found" while querying worker.log on secure cluster. The file actually exists."_

I probably have setup something wrong.

I used "mapredqa" (the nimbus admin account and also the topology submitter in my test) to login and I was able to view the log files on logviewer, i.e. no "page not found" issue. 
<img width="1436" alt="screen shot 2017-07-20 at 11 37 14 pm" src="https://user-images.githubusercontent.com/14900612/28449318-7429fa80-6da4-11e7-9d44-28091e19ff8f.png">

But if I use "ethan" account to login, I am able to see topology summary, component summary, except log files.  "ethan" is a nimbus.user, ui.user, logs.user and topology user of this topology. I got an warning in logviewer.log while querying for the worker.log

<img width="1440" alt="screen shot 2017-07-20 at 11 26 07 pm" src="https://user-images.githubusercontent.com/14900612/28449229-b9f8d69a-6da3-11e7-8141-5f95593d8b91.png">

I don't know the reason yet. I guess there could be something wrong in my cluster settings. It would be great if someone can test it on secure cluster about it.

Note that I was still testing the version before @revans2 's comments.

HeartSaVioR commented:
@Ethanlm 
Thanks for the update. I'll try to get it today and put my observation.
Btw, my additional commits didn't address anything about bug fix so it doesn't matter which commit you're testing with.

revans2 commented:
@HeartSaVioR I have not tested on a secure cluster at all.  It was all on a single node cluster.

I verified that as many links as I could worked correctly.  I verified I could switch between files, search through files.  I verified that log files were cleaned up as expected as the size grew.

The only other thing that I just noticed is that the pull down for selecting files does not have the current file selected as the default.  (again this can probably be done in a follow on JIRA)

HeartSaVioR commented:
@revans2 
I fixed the drop down issue, and also fixed issue regarding authorization check. I think it was the reason of issues what @Ethanlm observed.
@Ethanlm 
Could you test it again if you can spend time on this? If you don't have time to test, I'll just merge this after reviewing the latest commit from @revans2 .

Ethanlm commented:
@HeartSaVioR  I will test it quickly

Ethanlm commented:
Just tested (3). Don't know why (3) is still not working (even with nimbus admin account). 
![image](https://user-images.githubusercontent.com/14900612/28507762-00ad7c36-6ffb-11e7-92c0-1883b0a5f9f8.png)
(querying worker.log as nimbus admin works.)
![image](https://user-images.githubusercontent.com/14900612/28507778-114f6e3c-6ffb-11e7-8d5e-73a0e9f6eaf8.png)

But I think it's fine to fix it later if it's a real problem (if it doesn't only happen on my cluster) .

HeartSaVioR commented:
@Ethanlm 
Yeah I think I addressed (2) and forgot to debug (3). I'll test it quickly.

HeartSaVioR commented:
@Ethanlm 
I realized I missed addressing other places for parameter swapped. Just addressed and tested as well.
You can just build `storm-webapp` module and copy `storm-webapp jar` into `lib-webapp` directory.

Ethanlm commented:
Yea~ It works now. 
Use nimbus admin account: 
![image](https://user-images.githubusercontent.com/14900612/28508825-23b838ee-7003-11e7-81dc-42bd7e23191a.png)
Use "ethan" account (it's not authorized. Probably that why I got "page-not-found" while querying worker.log): 
![image](https://user-images.githubusercontent.com/14900612/28508831-298dbe38-7003-11e7-9014-97621e7ce83b.png)

A quick question.. "ethan" is nimbus.user, ui.user, and logs.user. But it's not able to view worker.log. Which configuration should I set "ethan" to..? Thanks very much.

HeartSaVioR commented:
@Ethanlm 
For now you should include `ethan` to `logs.users` in `topology configuration`. Global configuration doesn't take effect. This behavior seems same as current logviewer.
If we need to merge global configuration of `logs.users` and `logs.groups` into topology configuration, we can file a new issue and handle it accordingly.

Ethanlm commented:
@HeartSaVioR Thanks! It's good to learn

HeartSaVioR commented:
@Ethanlm 
No problem. Thanks for testing my patch thoughtfully. Please let me know if (2) becomes fixed with setting `logs.users` in the topology configuration.

Ethanlm commented:
Still not authorized.. But I guess it's my problem.. Will try to figure it out. Thanks.

![image](https://user-images.githubusercontent.com/14900612/28523809-8d57133e-7043-11e7-85cc-7cd72c94a198.png)
![image](https://user-images.githubusercontent.com/14900612/28523855-b8bdc266-7043-11e7-828b-b51e1de2bef8.png)

revans2 commented:
The change looks good to me and I am +1.  Will wait for @Ethanlm to work through his issues first though.

Ethanlm commented:
@HeartSaVioR @revans2  I think I figured it out. 
I added "ethan@DEREKD.YSTORM.NET" to logs.users and then it works. I traced down to the [isAuthorizedLogUser](https://github.com/apache/storm/pull/2204/files#diff-aad25f432acc108d98b3126bbd914873R74) function and found that it doesn't translate "principal" to "user" so that "ethan@DEREKD.YSTORM.NET" doesn't match with logs.user "ethan" in the configuration. @revans2 and I think the code is identical to the old code. So this should an existing bug.

revans2 commented:
@Ethanlm since you found this could you file a follow on JIRA to use the PrincipalToLocal plugin in the logviewer when authorizing users?

Ethanlm commented:
@HeartSaVioR  Yes. Sure. Should I wait until this PR being merged in?

HeartSaVioR commented:
@Ethanlm Great news for me to figure it out. I just added pair of `username` & `username@domain` while testing so didn't recognize the issue.
